File: test/reorder.test.ts
/**
 * Drag-to-reorder arithmetic.
 *
 *   npm run test:reorder
 *
 * Written after the fact, for a bug that shipped: dragging a room downwards
 * did nothing at all. The row you hold is drawn with a transform so it follows
 * your finger, which means its bounding box is always under the pointer — and
 * the hit test took the first row it found, which was that one. It compared
 * the result to where the drag started, saw no change, and returned.
 *
 * Upwards it happened to work, because the row above comes first in the array
 * and matched before the held row did. Exactly the kind of asymmetry nobody
 * spots by dragging something once and seeing it move.
 */

import { dropTarget, moveWithin, type Span } from '@/lib/reorder';

let failures = 0;

function check(label: string, ok: boolean, detail = '') {
  if (!ok) failures++;
  console.log(`${ok ? 'PASS' : 'FAIL'}  ${label}${detail ? `  — ${detail}` : ''}`);
}

/** Five rows, 50px tall, stacked from y=0. */
const ROWS: Span[] = [0, 1, 2, 3, 4].map((i) => ({ top: i * 50, bottom: i * 50 + 50 }));

/** The same, but row `held` has been dragged to sit over `y`. */
function dragging(held: number, y: number): Span[] {
  return ROWS.map((r, i) => (i === held ? { top: y - 25, bottom: y + 25 } : r));
}

function main() {
  /* ------------------------------------------------------------ the bug */

  // Holding row 1 and pulling down to the middle of row 2. The held row's box
  // has come with it and also covers the pointer.
  const down = dragging(1, 120);
  check('dragging down finds the row underneath', dropTarget(down, 120, 1) === 2, String(dropTarget(down, 120, 1)));
  check(
    'and not the row in your hand',
    dropTarget(down, 120, 1) !== 1,
    'the held row is under the pointer by construction',
  );

  // The old behaviour, for the record: without skipping the held row the
  // answer was the held row itself, which the caller read as "no change".
  const naive = down.findIndex((r) => 120 >= r.top && 120 <= r.bottom);
  check('the naive version returned the held row', naive === 1, `${naive}`);

  /* ------------------------------------------------------ the other way */

  const up = dragging(3, 80);
  check('dragging up finds the row above', dropTarget(up, 80, 3) === 1, String(dropTarget(up, 80, 3)));

  /* ------------------------------------------------------------- edges */

  check('a pointer past the end targets nothing', dropTarget(ROWS, 9999, 0) === null);
  check('a pointer above the list targets nothing', dropTarget(ROWS, -40, 4) === null);
  check('a gap between rows targets nothing', dropTarget([{ top: 0, bottom: 10 }, null], 40, 0) === null);
  check('an unmeasured row is skipped', dropTarget([null, { top: 0, bottom: 50 }], 25, 5) === 1);
  check('an empty list targets nothing', dropTarget([], 10, 0) === null);

  // Holding the first row and hovering it alone: nothing else to land on.
  check('the only row cannot be dropped on itself', dropTarget([ROWS[0]!], 25, 0) === null);

  /* ---------------------------------------------------------- the move */

  const list = ['a', 'b', 'c', 'd'];
  check('moving down', moveWithin(list, 0, 2).join('') === 'bcad', moveWithin(list, 0, 2).join(''));
  check('moving up', moveWithin(list, 3, 1).join('') === 'adbc', moveWithin(list, 3, 1).join(''));
  check('moving to the end', moveWithin(list, 0, 3).join('') === 'bcda');
  check('a move to the same place is the same list', moveWithin(list, 2, 2) === list);
  check('an out-of-range source is ignored', moveWithin(list, 9, 1) === list);
  check('an out-of-range target is ignored', moveWithin(list, 0, 9) === list);
  check('the original is not mutated', list.join('') === 'abcd');

  /* --------------------------------------------- a whole drag, in steps */

  // Pick up "a" and walk it to the bottom, one row at a time, the way the
  // component does: re-index after each swap, since the held row has moved.
  let order = ['a', 'b', 'c', 'd', 'e'];
  let from = 0;
  for (const target of [1, 2, 3, 4]) {
    const spans = dragging(from, target * 50 + 25);
    const at = dropTarget(spans, target * 50 + 25, from);
    check(`step to row ${target}`, at === target, String(at));
    order = moveWithin(order, from, at!);
    from = at!;
  }
  check('it ends up last', order.join('') === 'bcdea', order.join(''));

  console.log(failures === 0 ? '\nall green' : `\n${failures} failure(s)`);
  process.exit(failures === 0 ? 0 : 1);
}

main();

File: mobile/test/reorder_test.dart
/// Drag-to-reorder arithmetic.
///
///   dart test test/reorder_test.dart
///
/// Translated from `test/reorder.test.ts`. The bug this guards: the held row is
/// drawn following your finger, so it is always under the pointer by
/// construction. A hit test that takes the first match while dragging downwards
/// finds the held row, compares it to where it started, sees no change, and
/// does nothing. Dragging down never reordered anything.
library;

import 'package:stash_it/logic/reorder.dart';
import 'package:flutter_test/flutter_test.dart';

/// Five 50-pixel rows, stacked.
final rows = [
  const Span(0, 50),
  const Span(50, 100),
  const Span(100, 150),
  const Span(150, 200),
  const Span(200, 250),
];

void main() {
  group('the hit test', () {
    test('dragging down finds the row underneath', () {
      expect(dropTarget(rows, 120, 1), 2);
    });

    /*
      THE ASSERTION THE FILE EXISTS FOR. The held row is dragged down over its
      neighbour, so both boxes contain the pointer. Taking the first match
      returns the held row — index 1 — and the caller then does nothing at all.
    */
    test('and not the row in your hand, which is also under the pointer', () {
      final overlapping = [
        const Span(0, 50),
        const Span(100, 150), // held, dragged down onto row 2
        const Span(100, 150),
      ];
      final naive = overlapping.indexWhere((r) => 120 >= r.top && 120 <= r.bottom);
      expect(naive, 1, reason: 'the naive version returns the held row');
      expect(dropTarget(overlapping, 120, 1), 2);
    });

    test('dragging up finds the row above', () {
      expect(dropTarget(rows, 80, 3), 1);
    });

    test('a pointer past the end targets nothing', () {
      expect(dropTarget(rows, 9999, 0), isNull);
    });

    test('a pointer above the list targets nothing', () {
      expect(dropTarget(rows, -40, 4), isNull);
    });

    test('a gap between rows targets nothing', () {
      expect(dropTarget([const Span(0, 10), null], 40, 0), isNull);
    });

    // A null entry is a row that has not been measured yet, and is skipped
    // rather than treated as a zero-height row at the origin.
    test('an unmeasured row is skipped', () {
      expect(dropTarget([null, const Span(0, 50)], 25, 5), 1);
    });

    test('an empty list targets nothing', () {
      expect(dropTarget([], 10, 0), isNull);
    });

    test('the only row cannot be dropped on itself', () {
      expect(dropTarget([rows.first], 25, 0), isNull);
    });
  });

  group('the move', () {
    final list = ['a', 'b', 'c', 'd'];

    test('moving down', () => expect(moveWithin(list, 0, 2).join(), 'bcad'));
    test('moving up', () => expect(moveWithin(list, 3, 1).join(), 'adbc'));
    test('moving to the end', () => expect(moveWithin(list, 0, 3).join(), 'bcda'));

    // Identity, so a caller can use it to decide whether to rebuild.
    test('a move to the same place is the same list', () {
      expect(identical(moveWithin(list, 2, 2), list), isTrue);
    });

    test('an out-of-range source is ignored', () {
      expect(identical(moveWithin(list, 9, 1), list), isTrue);
      expect(identical(moveWithin(list, -1, 1), list), isTrue);
    });

    test('an out-of-range target is ignored', () {
      expect(identical(moveWithin(list, 0, 9), list), isTrue);
      expect(identical(moveWithin(list, 0, -1), list), isTrue);
    });

    test('the original is never mutated', () {
      moveWithin(list, 0, 3);
      expect(list.join(), 'abcd');
    });
  });
}
